batchglm/train/numpy/base_glm/estimator.py

```diff
@@ -86,8 +94,95 @@ def train_sequence(self, training_strategy, **kwargs):
+        theta_regularization: str = "od_factor",
+        use_geometric_mean: bool = True,
+        gmean_eps: float = 1.0,
+        bw_adjust: float = 1.5,
+    ):
+        logger.info("Performing Dispersion smoothing with flavour sctransform...")
+        # compute geometric log mean counts
+        if use_geometric_mean:
+            genes_log_gmean = np.log10(geometric_mean(self.model_container.x, axis=0, eps=gmean_eps))
+        else:
+            genes_log_gmean = np.log10(np.mean(self.model_container.x, axis=0))
+        if isinstance(genes_log_gmean, dask.array.core.Array):
+            genes_log_gmean = genes_log_gmean.compute()
+
+        # specify which kind of regularization is performed
+        scale_param = np.exp(self.model_container.theta_scale[0])  # TODO check if this is always correct
+        if theta_regularization == "log_theta":
+            dispersion_par = np.log10(scale_param)
+        elif theta_regularization == "od_factor":
+            dispersion_par = np.log10(1 + np.power(10, genes_log_gmean) / scale_param)
+        else:
+            raise ValueError(f"Unrecognized regularization method {theta_regularization}")
+        if isinstance(dispersion_par, dask.array.core.Array):
+            dispersion_par = dispersion_par.compute()
+
+        # downsample because KDE and bandwidth selection would take too long if performed on the entire dataset.
+        # It is sufficient to get a general idea of the data distribution
+        if len(genes_log_gmean) > 2000:
+            logger.info("Sampling 2000 random features...")
+            idx = np.random.choice(np.arange(len(genes_log_gmean)), 2000)
+            genes_log_gmean_filtered = genes_log_gmean[idx]
+            dispersion_par_filtered = dispersion_par[idx]
+        else:
+            dispersion_par_filtered = dispersion_par
+            genes_log_gmean_filtered = genes_log_gmean
+
+        # check for outliers in the function f(genes_log_gmean_filtered) = dispersion_par_filtered and remove them
+        logger.info("Searching for outliers...")
+        outliers = is_outlier(model_param=dispersion_par_filtered, means=genes_log_gmean_filtered)
+        outliers = outliers["outlier"].values
+        if np.any(outliers):
+            # toss out the outliers
+            logger.info(f"Excluded {np.sum(outliers)} outliers.")
+            genes_log_gmean_filtered = genes_log_gmean_filtered[~outliers]
+            dispersion_par_filtered = dispersion_par_filtered[~outliers]
+
+        # define a data grid with the downsampled and filtered values
+        domain_range = genes_log_gmean_filtered.min(), genes_log_gmean_filtered.max()
+        fd = FDataGrid(
+            data_matrix=dispersion_par_filtered, grid_points=genes_log_gmean_filtered, domain_range=domain_range
+        )
+        # select bandwidth to be used for smoothing
+        bw = bw_kde(genes_log_gmean_filtered) * bw_adjust
+        # bandwidth = FFTKDE(kernel='gaussian', bw='ISJ').fit(fd).bw * 0.37 * 3
```

Comment:
can we remove this comment or explain it?

Reply:
removed